v-next/hardhat/src/internal/builtin-plugins/coverage/hook-handlers/network.ts

```diff
@@ -0,0 +1,54 @@
+import type { HookContext, NetworkHooks } from "hardhat/types/hooks";
+import type { ChainType, NetworkConnection } from "hardhat/types/network";
+
+import { EdrProvider } from "../../network-manager/edr/edr-provider.js";
+
+export default async (): Promise<Partial<NetworkHooks>> => {
+  const handlers: Partial<NetworkHooks> = {
+    async newConnection<ChainTypeT extends ChainType | string>(
+      context: HookContext,
+      next: (
+        nextContext: HookContext,
+      ) => Promise<NetworkConnection<ChainTypeT>>,
+    ) {
+      const connection = await next(context);
+
+      if (context.globalOptions.coverage === true) {
+        if (connection.provider instanceof EdrProvider) {
+          const { getOrCreateCoverageManager } = await import(
+            "../coverage-manager.js"
+          );
```

Comment:
We access it via the getter so that we operate on an instance of the internal coverage manager type rather than the public one. I think technically we could force hast `context.coverage` to an internal type but it doesn't seem right either.

Reply:
This is the alternative I had in mind:

https://github.com/NomicFoundation/hardhat/compare/coverage-6581...coverage-6581-alt#diff-257ddac31ce1a21f944e89cd8a9575932b1d0a099a5f12a73aaac13f7b1589adR19-R21 

Actually, maybe it wouldn't be that bad. Especially, in comparison to this singleton.